### ISSUE TITLE
Jenkins log and artifact cleanup schedule

### ISSUE THREAD

**Issue by AdamBrousseau** [eclipse/openj9#1017]
Today we took down Eclipse's [hipp5](https://ci.eclipse.org/) which is where our OpenJ9 Jenkins master lives. We were consuming 264G space according to the [Bugzilla](https://bugs.eclipse.org/bugs/show_bug.cgi?id=530259) and have been asked to reduce.

What we store:
- Build logs (console output)
- SDKs from Builds and PRs

Jenkins has the following options, which are configured on a per job basis
```
Days to keep builds, if not empty, build records are only kept up to this number of days    
Max # of builds to keep, if not empty, only up to this number of build records are kept    
Days to keep artifacts, if not empty, artifacts from builds older than this number of days will be deleted, but the logs, history, reports, etc for the build will be kept    
Max # of builds to keep with artifacts, if not empty, only up to this number of builds have their artifacts retained
```
Currently, our nightly (and OMR acceptance) builds are setup with the following:
- Days to keep builds: 60
- Max # of builds to keep: 100
- Days to keep artifacts:
- Max # of builds to keep with artifacts: 10

Our PR builds are not configured to discard (keep everything).

#### Propose:
Nightlies and OMR Acceptance:
- Days to keep builds: 60
- Max # of builds to keep: 100
- Days to keep artifacts: ~7~
- Max # of builds to keep with artifacts: ~10~

PRs:
- Days to keep builds: 60
- Max # of builds to keep: 100
- Days to keep artifacts: ~14~
- Max # of builds to keep with artifacts: ~25~

Please review @pshipton 

Also, I have a script to check disk space I could add a job for and send a Slack notification if we're over X. Could open as a separate issue if we're interested.

**Comment by pshipton** [comment]
I'm ok with this, or with smaller numbers even. Do you know about how many builds consume the 264G?

Perhaps others will have comments.

**Comment by pshipton** [comment]
I checked a full Java 9 build and it was 4G uncompressed. This includes the git repos, all the source, build artifacts, and images. I assume the SDK kept around must be smaller than this. The jdk image itself is 323M and zipped its 202M, so around 20G for 100 builds.

**Comment by pshipton** [comment]
We probably want a bigger number for PRs, "Max # of builds to keep with artifacts", say 100 to match the "Max # of builds to keep".

**Comment by AdamBrousseau** [comment]
Biggest offenders currently are the PR Sanity builds. (no cleanuo yet)
688 SDKs being stored across all jobs.
Java 9 SDKs are ~200M and 8's are ~100M (tar.gz)

**Comment by AdamBrousseau** [comment]
Based on how much space we've consumed in ~3 months and suggestions in comments, I have updated the numbers in the description. Propose we keep 2 months/100 builds?

**Comment by DanHeidinga** [comment]
@AdamBrousseau Are there any download stats on how often these artifacts have been downloaded?  If they aren't being downloaded, then storing them serves no purpose.

**Comment by pshipton** [comment]
I guess a larger number of days to keep builds doesn't much matter if they are limited to 100 builds. 

What does it mean not to put a limit on the artifacts? Won't the artifacts eventually consume too much space, or do they disappear with the builds?

**Comment by smlambert** [comment]
Artifacts should disappear with the builds (from Jenkins master).  

Based on typical use cases, do we realistically need to keep the SDK for 100 builds for PR builds?  I had thought the use case was, "as a developer I want to download the SDK from my PR build, immediately after I build it, so I can do some additional local testing.  After a few days (x number of builds later), I want to rebuild, as I want to keep current with all of the other changes coming into the repo)."  As a developer, do I ever want to use an SDK that is 100 builds old?

**Comment by AdamBrousseau** [comment]
> What does it mean not to put a limit on the artifacts?

The artifacts would get cleaned up with the build logs.

> download stats

I don't see anything built into Jenkins for this. Initial searching only talks about website monitoring services.

**Comment by pshipton** [comment]
Are you saying we get 50 builds per job? i.e. 50 for Java 9 pLinux, 50 for Java 9 zLinux, etc.

**Comment by pshipton** [comment]
I'd say make the change, and then we can adjust later as necessary.

**Comment by AdamBrousseau** [comment]
Thought about it a bit more. Makes more sense to me to do something like this
- Days to keep builds:
- Max # of builds to keep: 50
- Days to keep artifacts: 60
- Max # of builds to keep with artifacts:

Modified all build configs.

**Comment by AdamBrousseau** [comment]
Reopening this for discussion as we've got an eclipse bugzilla opened against us for consuming 365G space. They would like us to stay under 100.
https://bugs.eclipse.org/bugs/show_bug.cgi?id=533823
Propose we add `Max # of builds to keep with artifacts:` set to `X` for PR builds and `Y` for Build jobs.
Currently an example PR job that is keeping 50 builds and 60 days for artifacts. There are 50 artifacts (all builds are less than 60 days old) and its consuming ~13G space (50*205M).

Some napkin math:
- JDK8,9,10: `100M + 200M + 200M = 500M`
- x (build job + PR(Compile, Sanity Extended) repos: 4 (ignoring PRs from the extensions repos as we've had less than 10 builds total per PR build, otherwise this number would be 7)
- = 2G per platform for 1 of each build
Saving 20 SDKs for PRs and 10 for Build jobs
- = 35G per platform
Currently we have 3 platforms
- = ~105G
When we add Windows next week we would be up to 140G.
If we reduce to last 10 PRs and last 10 builds
- = ~60G + ~20G when we add Windows

To summarize, Propose we keep the last 10 of every build (PullRequest* or Build*).

Disclaimer: This does not account for build logs
- ~20M per Build job or PR Compile * 18
- ~100M per PR Sanity * 9
- ~30M per PR Extended * 9
- ~70M per Test Sanity * 9
- ~15M per Test Extended * 9
- = 2.2G * 50 (builds to keep) = ~115G
^ I think these numbers are worst case scenario. I think we should worry about the artifacts for now and see where we stabilize over the next week or 2.

I also have a job I can setup to monitor the space we are consuming and list the top 10 "Pig" jobs. Optional slack notifications as well.

**Comment by DanHeidinga** [comment]
Agreed.  Though I'm fine with decreasing the number of builds kept until there is evidence someone is actually downloading them.

**Comment by AdamBrousseau** [comment]
Updated all the build and PR jobs to only keep last 10 artifacts. I will keep an eye on the disk space over the next few days to see if it reduces.

**Comment by AdamBrousseau** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=535057
Email from Eclipse:
```
Hi,
We've been facing a serious outage on the machine which runs your Jenkins instance today.
There were no more space left on disk. It appears that OpenJ9 is using 252GB of disk space for its
jobs (out of the 1TB available which is shared by all 17 projects which have a Jenkins on the
same machine). Could you please do some housekeeping?
Thanks.

--
Mikaël Barbero - Eclipse Foundation
IT Services - Release Engineering
📱 (+33) 642 028 039
📧 mikael.barbero@eclipse-foundation.org
```

**Comment by AdamBrousseau** [comment]
Joe delivered a change to stop archiving SDKs in PR builds (#1988)
Our usage has gone from 229G to 175G.
As discussed in Slack, I will reconfigure the other jobs to only store 5 SDKs instead of 10. I will post back once that change has propagated through.
```

**Comment by AdamBrousseau** [comment]
Only storing 5 SDKs per build now. Usage went from 180G to 173G.
Also noticed that the workspaces on master are consuming 35G. These are git repos from when we do the initial checkouts before we decide which node to run on. It should be easy to ensure these are cleaned up when we leave the node. I can also add a ref repo to the master to save clone times. I'd like to block this change on #1897.

**Comment by pshipton** [comment]
What was the target size Eclipse suggested? I've got 100G in my head but I might be confusing Eclipse ci space with something else.

**Comment by AdamBrousseau** [comment]
100G

**Comment by pshipton** [comment]
While waiting for #1897 it seems we have no other choice than to reduce the number of logs. We could try 40 instead of 50 to see how the space usage changes.

**Comment by AdamBrousseau** [comment]
Cleaned up all the workspaces on master. Reduced usage from 185G to 148G. Will keep an eye on these and look into a solution for not keeping them around.

**Comment by AdamBrousseau** [comment]
>We could try 40 instead of 50

Seems the Sanity logs are larger than the Extended or Compile. Do we want to reduce to 40 across the board or something else? Only for PR sanity, all sanity, just compiles?

**Comment by pshipton** [comment]
Lets try 40 across the board and see where that puts us for space.

**Comment by pshipton** [comment]
Can also reduce non-PR sanity/extended/compile to 10 builds, as I expect failures in nightly or omr acceptance will get looked at quickly.

**Comment by smlambert** [comment]
@AdamBrousseau - we will update the nightlies to keep last 10, based on a discussion in releng interlock today.  So don't worry about that, it will be taken care of for nightly builds and tests.

**Comment by pshipton** [comment]
All pipeline jobs (https://ci.eclipse.org/openj9/view/Pipelines/) have been set to keep 10 builds.

**Comment by llxia** [comment]
All test jobs have been set to keep 10 builds https://ci.eclipse.org/openj9/view/Test/

**Comment by AdamBrousseau** [comment]
Did someone update Build*?

Two things:
- I recently learned this setting can be moved into our pipeline code
```
properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '5', daysToKeepStr: '', numToKeepStr: '10'))])
```
- I discovered the [Compress Build Log Plugin](https://wiki.jenkins.io/display/JENKINS/Compress+Build+Log+Plugin) which is supposed to compress the logs. When I tried it out it created a log.gz but did not cleanup the uncompressed logs. Haven't had a chance to debug this yet so have not turned it on in the builds.

**Comment by smlambert** [comment]
We talked about whether you wanted that buildDiscarder in the pipeline code or not.  
Since its not there yet, we manually changed some of the nightlies.  Pipeline settings override the manual settings, so if/when its incorporated into pipeline code, should be no biggie.

No one has done the Build* yet.

**Comment by pshipton** [comment]
Build, only for Build-JDK* is done now https://ci.eclipse.org/openj9/view/Build/

**Comment by AdamBrousseau** [comment]
Usage down to 123G

**Comment by AdamBrousseau** [comment]
Crept back up to 164G

"Pig" jobs:
- 5.4G	PullRequest-Sanity-JDK10-linux_x86-64-OpenJ9
- 5.4G	PullRequest-Sanity-JDK10-linux_x86-64_cmprssptrs-OpenJ9
- 5.2G	PullRequest-Sanity-JDK10-aix_ppc-64_cmprssptrs-OpenJ9
- 4.8G	PullRequest-Sanity-JDK10-win_x86-64_cmprssptrs-OpenJ9
- 4.6G	PullRequest-Sanity-JDK8-linux_x86-64-OpenJ9
- 4.5G	PullRequest-Sanity-JDK8-linux_390-64_cmprssptrs-OpenJ9
- 4.5G	PullRequest-Sanity-JDK10-linux_390-64_cmprssptrs-OpenJ9
- 4.4G	PullRequest-Sanity-JDK10-linux_ppc-64_cmprssptrs_le-OpenJ9
- 4.1G	PullRequest-Sanity-JDK8-aix_ppc-64_cmprssptrs-OpenJ9
- 4.0G	PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9
- 3.9G	PullRequest-Sanity-JDK9-linux_ppc-64_cmprssptrs_le-OpenJ9
- 3.8G	PullRequest-Sanity-JDK9-linux_390-64_cmprssptrs-OpenJ9
- 3.7G	PullRequest-Sanity-JDK8-win_x86-64_cmprssptrs-OpenJ9
- 3.6G	PullRequest-Sanity-JDK9-aix_ppc-64_cmprssptrs-OpenJ9
- 3.5G	PullRequest-Sanity-JDK9-linux_x86-64_cmprssptrs-OpenJ9
- 3.5G	PullRequest-Sanity-JDK8-linux_ppc-64_cmprssptrs_le-OpenJ9
- 3.3G	PullRequest-Sanity-JDK8-win_x86-OpenJ9
- 3.0G	PullRequest-Extended-JDK8-aix_ppc-64_cmprssptrs-OpenJ9
- 2.6G	PullRequest-Sanity-JDK9-linux_x86-64-OpenJ9
- 2.6G	PullRequest-Extended-JDK10-linux_ppc-64_cmprssptrs_le-OpenJ9

Basically comes down to 100M worth of test logs x50 builds of history = 5G

**Comment by pshipton** [comment]
Shouldn't there be some jdk11 jobs in the list now?

**Comment by AdamBrousseau** [comment]
We haven't had enough 11 builds to use that amount of space

**Comment by AdamBrousseau** [comment]
Can this be closed and reopened later if needed?

Since the last comment, we've added Artifactory, which moved all our binaries off Master. We've also changed the way our builds are arranged (#2836 and #5182) and how many we store of each (#6021).
